apps/vast/pyvast_threatbus/pyvast_threatbus.py

```diff
@@ -566,53 +566,90 @@ async def report_sightings(
-                ser = json.loads(sighting.serialize())
-                ser[
-                    ThreatBusSTIX2Constants.X_THREATBUS_SIGHTING_CONTEXT.value
-                ] = transformed_context
-                sighting = parse(json.dumps(ser), allow_custom=True)
-            except Exception as e:
-                logger.error(
-                    f"Cannot parse transformed sighting context (expecting JSON): {transformed_context_raw}",
-                    e,
+            if not context:
+                logger.warn(
+                    f"Cannot report sighting context to custom sink because no context data is found in the sighting {sighting}"
                 )
                 continue
-        if (
-            sink
-            and ThreatBusSTIX2Constants.X_THREATBUS_SIGHTING_CONTEXT.value
-            in sighting.object_properties()
-        ):
             if sink.lower() == "stdout":
-                print(json.dumps(sighting.x_threatbus_sighting_context))
+                print(json.dumps(context))
             else:
-                await invoke_cmd_for_context(
-                    sink, sighting.x_threatbus_sighting_context
-                )
+                await invoke_cmd_for_context(sink, context)
         else:
             socket.send_string(f"{topic} {sighting.serialize()}")
         sightings_queue.task_done()
         logger.debug(f"Reported sighting: {sighting}")
 
 
+async def transform_context(sighting: Sighting, transform_cmd: str) -> Sighting:
+    """
+    Transforms the context of a sighting using the command configured in
+    `transform_context`
+    @param sighting the sighting as it was reported by VAST
+    @param transform_cmd The command to use to pipe sightings to. Treated
+        as template string: occurrences of '%ioc' in the cmd string get replaced
+        with the matched IoC.
+    @return a copy of the original sighting with the x_threatbus_context field
+        set and transformed accordingly
+    """
+    context = (
+        sighting.x_threatbus_sighting_context
+        if ThreatBusSTIX2Constants.X_THREATBUS_SIGHTING_CONTEXT.value
+        in sighting.object_properties()
+        else None
+    )
+    if not context:
+        logger.error(
+            f"Cannot invoke `transform_context` command because no context data is found in the sighting {sighting}"
+        )
+        return
+    indicator = (
+        sighting.x_threatbus_indicator
+        if ThreatBusSTIX2Constants.X_THREATBUS_INDICATOR.value
+        in sighting.object_properties()
+        else None
+    )
+    if indicator:
+        _, ioc_value = split_object_path_and_value(indicator.pattern)
+    else:
+        # try to find the indicator value instead
+        ioc_value = (
+            sighting.x_threatbus_indicator_value
+            if ThreatBusSTIX2Constants.X_THREATBUS_INDICATOR_VALUE.value
+            in sighting.object_properties()
+            else None
+        )
+    if not ioc_value:
```

Comment:
I don't understand how this works. `ioc_value` has not been declared in this scope.

Reply:
Scoping in Python is different than in C++. e.g., `if` does not scope. Instead, variables declared inside an `if`, `while` etc have the scope of the enclosing function. See LEBG scope (google) or this [SO answer](https://stackoverflow.com/a/2829642/2525184)

Reply:
TIL. :confused: